### ISSUE TITLE
str.maketrans false error

### ISSUE THREAD

**Issue by brianmay** [python/mypy#4374]
From https://github.com/python/typeshed/issues/1788:

The full original source was:

```
mac_value = int(mac.translate(str.maketrans(dict([(x, None) for x in [" ", ".", ":", "-"]]))), 16)
```

This gives the error:

```
aiolifxc/aiolifx.py:61: error: Argument 1 to "maketrans" of "str" has incompatible type "Dict[str, None]"; expected "Union[Dict[int, Union[int, str, None]], Dict[str, Union[int, str, None]], Dict[Union[str, int], Union[int, str, None]]]"
```

This is nonsense. ``Dict[str, None]`` is clearly allowed by the union type (2nd part) that is given: ``Dict[str, Union[int, str, None]]``

When I break this down into two lines:

```
trans = str.maketrans(dict([(x, None) for x in [" ", ".", ":", "-"]]))
mac_value = int(mac.translate(trans), 16)
```

mypy is now happy. As such, I suspect mypy has some sort of bug dealing with the complicated statement.

**Comment by JelleZijlstra** [comment]
I'm not sure whether there's a bug here and whether it's in mypy or typeshed, but a few important observations:

- The relevant type declaration is at https://github.com/python/typeshed/blob/master/stdlib/3/builtins.pyi#L286: `def maketrans(x: Union[Dict[int, _T], Dict[str, _T], Dict[Union[str, int], _T]]) -> Dict[int, _T]: ...`, where `_T = TypeVar('_T')`.
- `Dict[str, None]` is actually *not* allowed by `Dict[str, Union[int, str, None]]`, because `Dict` is invariant. See http://mypy.readthedocs.io/en/latest/common_issues.html#invariance-vs-covariance.
- I'm not sure where mypy gets the union types in the dict value fields from. Something weird with overloads and typevars?

I can reproduce the OP's problem with:
```
$ cat maketrans.py 
mac = ''
mac_value = int(mac.translate(str.maketrans(dict([(x, None) for x in [" ", ".", ":", "-"]]))), 16)
$ python -m mypy --strict-optional maketrans.py 
maketrans.py:2: error: Argument 1 to "maketrans" of "str" has incompatible type "Dict[str, None]"; expected "Union[Dict[int, Union[int, str, None]], Dict[str, Union[int, str, None]], Dict[Union[str, int], Union[int, str, None]]]"
```

**Comment by gvanrossum** [comment]
maketrans() is a dinosaur. The one-arg form really requires a dict (can't
be another mapping), the keys are Union[int, str] and the values are
Union[int, str, None]. If you write that as Dict[Union[int, str],
Union[int, str, None]] then it's fiendishly hard to come up with a
validly-typed argument because dict is invariant. So typeshed tries to
expand the unions, but it doesn't go far enough. The full expansion would
be 2 key types times 7 value types, IOW a union of 14 types. There are 7
possible value types because it can be int, str, None, any two of those,
and all three.

**Comment by AlexWaygood** [comment]
Reproduced on 0.941, but now we get a three-line error message instead of a one-liner (hooray for verbosity?)

```
main.py:2: error: Argument 1 to "maketrans" of "str" has incompatible type "Dict[str, None]"; expected "Union[Dict[int, Union[int, str, None]], Dict[str, Union[int, str, None]], Dict[Union[str, int], Union[int, str, None]]]"
main.py:2: note: "Dict" is invariant -- see https://mypy.readthedocs.io/en/stable/common_issues.html#variance
main.py:2: note: Consider using "Mapping" instead, which is covariant in the value type
```

**Comment by AlexWaygood** [comment]
Mypy also [seems to be happy](https://mypy-play.net/?mypy=latest&python=3.10&flags=strict&gist=e146a726343f4e488aaddb0eed6e9058) if you use `dict.fromkeys` instead of using a list comprehension inside the `dict()` constructor. And that's more idiomatic Python anyway:

```python
mac=''
mac_value = int(mac.translate(str.maketrans(dict.fromkeys([" ", ".", ":", "-"]))), 16)
```

So, given the easy workarounds, I'd say this is pretty low priority.

**Comment by hauntsaninja** [comment]
mypy's new error message makes it clearer what the issue is. Unfortunately, maketrans seems to require exactly a dict type, not an arbitrary mapping.
